Dev/kwiver integration

linus-sherrill opened the pull request.

linus-sherrill committed:
Updated Doxygen config file to newer version

linus-sherrill committed:
Updated for newer doxygen program

Refactored config file to make it easier to upgrade doxygen

linus-sherrill committed:
Add target to build combined documentation

linus-sherrill committed:
Documentation and error message improvements.

linus-sherrill committed:
Documentation changes.

linus-sherrill committed:
Improve documentation

linus-sherrill committed:
Add config output operator to simplify debugging.

linus-sherrill committed:
Added logging and improved documentation

linus-sherrill committed:
Improved documentation, added examples

linus-sherrill committed:
Expanding explanations

kfieldho committed:
Add missing include for release builds

linus-sherrill committed:
Improve documentation

linus-sherrill committed:
Additional documentation

linus-sherrill committed:
Add program description to tools infrastructure.

There was no way of describing what a tool does or why I would want to
use it (even though the tools had really cool names).

Tool help should be equivalent to a man page. It's not that complete
yet, but at least there is a way to add more description.

linus-sherrill committed:
Add program description to tools infrastructure.

There was no way of describing what a tool does or why I would want to
use it (even though the tools had really cool names).

Tool help should be equivalent to a man page. It's not that complete
yet, but at least there is a way to add more description.

linus-sherrill committed:
Add logging

Well it is not really logging, but it gets the job done.

linus-sherrill committed:
Check for null pointer.

Sometimes there is a null pointer for the config.
Problem discovered by test routines.

mathstuf commented on README.md:
```diff
@@ -69,7 +69,7 @@ C++
   * Use braces around all control (even single-line if) blocks
   * Use typedefs
-  * Use exceptions and return values, not error codes and input parameters
+  * Use exceptions and return values, not error codes and output parameters
```
I merged Kyle's branch from #1 for this.

linus-sherrill replied:
I don't see this on master. Did it not get pushed or is it on another
branch?

On Fri, Aug 22, 2014 at 2:17 PM, Ben Boeckel notifications@github.com
wrote:

> In README.md:
> 
> > @@ -69,7 +69,7 @@ C++
> > - Export symbols (or import them if possible)
> > - Use braces around all control (even single-line if) blocks
> > - Use typedefs
> >   -  \* Use exceptions and return values, not error codes and input parameters
> >   -  \* Use exceptions and return values, not error codes and output parameters
> 
> I merged Kyle's branch from #1 https://github.com/Kitware/sprokit/pull/1
> for this.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Kitware/sprokit/pull/2/files#r16612608.

## 

*Linus Sherrill - *R&D Engineer
Kitware, Inc.
28 Corporate Drive
Clifton Park, NY 12065-8662
E: linus.sherrill@kitware.com
P: 518.881.4400

mathstuf replied:
On Mon, Aug 25, 2014 at 09:01:47 -0700, linus-sherrill wrote:

> I don't see this on master. Did it not get pushed or is it on another branch?

Oops; didn't push. Done now.

mathstuf commented on extra/doxygen/Doxyfile.in:
```diff
@@ -1,109 +1,1798 @@
-# Doxyfile 1.5.7.1-KDevelop
+# Doxyfile 1.8.1.1
```
Could the file be cleaned of the comments to match the old version closer? It's near impossible to review as-is.

mathstuf commented on src/CMakeLists.txt:
```diff
@@ -13,3 +13,10 @@ endif ()
+
+sprokit_create_doxygen(
+  "${CMAKE_CURRENT_SOURCE_DIR}"
+  sprokit-master)
```
Maybe `sprokit-all` (`master` is too reminiscent of the branch name)?

mathstuf commented on src/sprokit/pipeline/config.cxx:
```diff
@@ -400,4 +400,22 @@ strip_block_name(config::key_t const& subblock, config::key_t const& key)
 }
 
+
+std::ostream& operator<<( std::ostream& str, sprokit::config const& obj )
```
Bad whitespace (extra line before it, spaces around `<<` and excess spaces in the parens).

mathstuf commented on src/sprokit/pipeline/config.cxx:
```diff
@@ -400,4 +400,22 @@ strip_block_name(config::key_t const& subblock, config::key_t const& key)
+    std::string ro;
+
+    sprokit::config::value_t const val = obj.get_value< sprokit::config::value_t > ( key );
+    if (obj.is_read_only( key ) ) ro = "[ro]";
```
Missing braces.

mathstuf commented on src/sprokit/pipeline/config.h:
```diff
@@ -43,11 +43,46 @@
 #include <cstddef>
 
+
+#include <cxxabi.h>
```
This stuff belongs in a source file (probably `utils.cxx`). I really don't like having `ifdef`fery in headers (ABI insanity will follow).

linus-sherrill replied:
Unfortunately the code needs to go in header file since it is templated
over an unknown set of types. It is informative to report destination type
when a lexical_cast fails but is there another way of doing this?
This code gave me the best results during debugging. It is only enabled
under linux because I could not test it on windows (maybe it works there
too?)

On Fri, Aug 22, 2014 at 2:22 PM, Ben Boeckel notifications@github.com
wrote:

> In src/sprokit/pipeline/config.h:
> 
> >  #include <typeinfo>
> >  #include <vector>
> > 
> >  #include <cstddef>
> > 
> > +
> > +#include <cxxabi.h>
> 
> This stuff belongs in a source file (probably utils.cxx). I really don't
> like having ifdeffery in headers (ABI insanity will follow).
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Kitware/sprokit/pull/2/files#r16612949.

## 

*Linus Sherrill - *R&D Engineer
Kitware, Inc.
28 Corporate Drive
Clifton Park, NY 12065-8662
E: linus.sherrill@kitware.com
P: 518.881.4400

mathstuf replied:
On Mon, Aug 25, 2014 at 09:33:21 -0700, linus-sherrill wrote:

> Unfortunately the code needs to go in header file since it is templated over an
> unknown set of types. It is informative to report destination type when a
> lexical_cast fails but is there another way of doing this? This code gave me
> the best results during debugging. It is only enabled under linux because I
> could not test it on windows (maybe it works there too?)

The typeid(T) has to be in a header, but the demangling doesn't have to
be. Adding a _demangle_symbol(char const*) function (or similar) which
is in a source file and abstracted out. The default implementation would
just return the input string.

mathstuf commented on src/sprokit/pipeline/config.h:
```diff
@@ -119,8 +154,8 @@ class SPROKIT_PIPELINE_EXPORT config
-     * \throws no_such_configuration_value_exception Thrown if the requested index does not exist.
-     * \throws bad_configuration_cast_exception Thrown if the cast fails.
+     * \throw no_such_configuration_value_exception Thrown if the requested index does not exist.
+     * \throw bad_configuration_cast_exception Thrown if the cast fails.
```
Were all the `throws` → `throw` changes necessary? [Doxygen](http://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdthrow) says the two are equivalent.

mathstuf commented on src/sprokit/pipeline/config.h:
```diff
@@ -212,7 +249,7 @@ class SPROKIT_PIPELINE_EXPORT config
      * \brief Return the values available in the configuration.
      *
-     * \returns All of the keys available within the block.
+     * \return All of the keys available within the block.
```
Same for all the `returns` → `return` changes.

mathstuf commented on src/sprokit/pipeline/config.h:
```diff
@@ -430,10 +469,14 @@ config_cast_default(config::value_t const& value)
+    std::stringstream ss;
+    ss << e.what() << " - converting \"" << value << "\" to type \""
+       << type_name<T>() << "\"";
+    throw bad_configuration_cast( ss.str().c_str() );
```
No need for the `c_str` call here (or there shouldn't be at least).

linus-sherrill replied:
The only constructor for this exception is
    bad_configuration_cast(char const\* reason) SPROKIT_NOTHROW;

It is possible that this is not the best exception to throw here. An
alternative is to use the "bad_configuration_cast_exception", but there is
no "key" available at this point.

On Fri, Aug 22, 2014 at 2:25 PM, Ben Boeckel notifications@github.com
wrote:

> In src/sprokit/pipeline/config.h:
> 
> > @@ -430,10 +469,14 @@ config_cast_default(config::value_t const& value)
> >    }
> >    catch (boost::bad_lexical_cast const& e)
> >    {
> > -    throw bad_configuration_cast(e.what());
> > -    std::stringstream ss;
> > -    ss << e.what() << " - converting \"" << value << "\" to type \""
> > -       << type_name<T>() << "\"";
> > -    throw bad_configuration_cast( ss.str().c_str() );
> 
> No need for the c_str call here (or there shouldn't be at least).
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Kitware/sprokit/pull/2/files#r16613117.

## 

*Linus Sherrill - *R&D Engineer
Kitware, Inc.
28 Corporate Drive
Clifton Park, NY 12065-8662
E: linus.sherrill@kitware.com
P: 518.881.4400

mathstuf replied:
On Mon, Aug 25, 2014 at 10:07:08 -0700, linus-sherrill wrote:

> The only constructor for this exception is bad_configuration_cast(char const*
> reason) SPROKIT_NOTHROW; It is possible that this is not the best exception to
> throw here. An alternative is to use the "bad_configuration_cast_exception",
> but there is no "key" available at this point.

Oops, should have seen that it didn't end with '_exception' and
therefore wasn't a user-seen exception. Above, it is caught and the key
information is added. The intermediate storage for the string is a
std::string (m_what), so this is fine.

mathstuf commented on src/sprokit/pipeline/config.h:
```diff
@@ -451,6 +494,7 @@ config_cast_inner(config::value_t const& value)
   return config_cast_default<T>(value);
 }
 
+
```
Excess whitespace (same sprinkled around).

mathstuf commented on src/sprokit/pipeline/config.h:
```diff
@@ -520,6 +566,8 @@ ::get_value(key_t const& key, T const& def) const SPROKIT_NOTHROW
   }
 }
 
+std::ostream& operator<<(std::ostream&str, sprokit::config const& obj);
```
I'd rather this be a member if possible. Also whitespacing.

mathstuf commented on src/sprokit/pipeline/edge.cxx:
```diff
@@ -128,6 +128,12 @@ ::edge(config_t const& config)
 
+  if ( capacity != 0 )
+  {
+    /// \todo LOG_DEBUG - Need an edge name like from.type.to
```
This can be added to the config block by the pipeline.

mathstuf commented on src/sprokit/pipeline/pipeline.cxx:
```diff
@@ -1066,6 +1066,11 @@ ::priv(pipeline* pipe, config_t conf)
   , setup_successful(false)
   , running(false)
 {
+  if ( config ) // sometimes config may be null
```
`config` should be checked for `NULL` in the `pipeline` ctor before creating the `priv` object.

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -116,9 +121,18 @@ class SPROKIT_PIPELINE_EXPORT process
+    /// 3/1). Frequencies less than one should have a numerator of one
+    /// (e.g. 1/3, 1/5). Frequencies that are not of the above form,
+    /// but do resolve to one of those forms are not supported
+    /// (e.g. 3/3, 2/4. 4/2).
```
Why would these not be supported? If they're not, maybe a `size_t` and an up/down enum would fit better?

linus-sherrill replied:
I could not find any documentation on the frequency concept, so I tried to
reverse engineer from the code. Apparently I missed some of the details :-(
 What is the desired behaviour of frequency or how is the best way to
describe this type?

On Fri, Aug 22, 2014 at 2:29 PM, Ben Boeckel notifications@github.com
wrote:

> In src/sprokit/pipeline/process.h:
> 
> > @@ -116,9 +121,18 @@ class SPROKIT_PIPELINE_EXPORT process
> >      typedef std::string port_type_t;
> >      /// The type for the component of a frequency.
> >      typedef size_t frequency_component_t;
> > -    /// The type for the frequency of data on a port.
> >   +
> > -    /// \brief The type for the frequency of data on a port. Since
> > -    /// this is a rational type, the frequency can be more or less
> > -    /// than one. Frequencies above one should be integers (e.g. 1/1,
> > -    /// 3/1). Frequencies less than one should have a numerator of one
> > -    /// (e.g. 1/3, 1/5). Frequencies that are not of the above form,
> > -    /// but do resolve to one of those forms are not supported
> > -    /// (e.g. 3/3, 2/4. 4/2).
> 
> Why would these not be supported? If they're not, maybe a size_t and an
> up/down enum would fit better?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Kitware/sprokit/pull/2/files#r16613365.

## 

*Linus Sherrill - *R&D Engineer
Kitware, Inc.
28 Corporate Drive
Clifton Park, NY 12065-8662
E: linus.sherrill@kitware.com
P: 518.881.4400

mathstuf replied:
Yeah, the docs are sparse mainly because the tests for them are also sparse (i.e., their behavior is not nailed down by the tests). Basically a process has a single "step" which is 1:1. If one port is pulled twice and another thrice, it doesn't matter if one is 2:1 and the other 3:1 or one 2:3 and the other 1:1. The pipeline takes all of the port frequencies and determines a common frequency for the entire pipeline then tells processes about it with `set_core_frequency` so that all vales are whole numbers which is also the stamp step value expected/generated on the port automatically. So in the end, they will all be whole numbers, but the ratio of them to each other is all that matters within a single process since the pipeline will figure out a global multiplier for all process the same anyways (modulo overflowing the `uint64_t` stamp counter which I hope isn't common).

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -145,7 +159,9 @@ class SPROKIT_PIPELINE_EXPORT process
          * \param description_ A description of the port.
-         * \param frequency_ The frequency of the port relative to the step.
+         * \param frequency_ The frequency of the port relative to the
+         * step.  See process::set_output_port_frequency() or
```
Use a `\sa` directive?

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -116,9 +121,18 @@ class SPROKIT_PIPELINE_EXPORT process
     typedef boost::rational<frequency_component_t> port_frequency_t;
+
     /// The type for a flag on a port.
+    ///@todo Add reference to predefined port flags.
```
Use `\`, not `@`.

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -236,6 +252,7 @@ class SPROKIT_PIPELINE_EXPORT process
      * \note This is only exposed for easier access from bindings.
      *
      * All levels include lower levels.
+     * \sa set_data_checking_level
```
Extra space above?

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -383,75 +400,82 @@ class SPROKIT_PIPELINE_EXPORT process
-     * \throws static_type_reset_exception Thrown when the \p port's current type is not dependent on other types.
-     * \throws set_type_on_initialized_process_exception Thrown when the port type is set after initialization.
+     * \throw no_such_port_exception Thrown when \p port does not
+     * exist on the process.  \throw static_type_reset_exception
```
Second `\throw` directive should start on its own line.

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -383,75 +400,82 @@ class SPROKIT_PIPELINE_EXPORT process
+     * exist on the process.  \throw static_type_reset_exception
+     * Thrown when the \p port's current type is not dependent on
+     * other types.
+     *
```
Extra line.

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -463,32 +487,125 @@ class SPROKIT_PIPELINE_EXPORT process
-    /// A type which indicates that the type depends on the connected port's type.
+
+    /**
+     *  \brief A type which indicates that the type depends on the
```
Extra indentation.

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -463,32 +487,125 @@ class SPROKIT_PIPELINE_EXPORT process
+     *
+     * The process can determine the type, but it must be configured
+     * before the type can be pinned down. The is usually for
+     * processes which read data from a file which it may not know
```
read → reads

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -463,32 +487,125 @@ class SPROKIT_PIPELINE_EXPORT process
+     * used when data is just passing through a process and the actual
+     * type is not known.
+     *
+     *  If a tag is appended to the type, then when any of the ports
```
Extra indentation.

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -463,32 +487,125 @@ class SPROKIT_PIPELINE_EXPORT process
+    /**
+     * \brief A flag which indicates that the output is shared between
+     * receivers.
+     *
```
More information could certainly belong here. `shared` is probably the most complex flag currently.

mathstuf replied:
For the extra information: a `shared` port may be connected to exactly one `mutable` port, any number of non-mutable ports, or nothing. Any other usage is a data sharing violation and not allowed. One TODO item is that port flags are not propagated across flow-dependent ports (more API is required here).

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -463,32 +487,125 @@ class SPROKIT_PIPELINE_EXPORT process
+     *
+     * If this port is not connected, the value supplied is taken from
+     * a specific config entry. The config entry is automatically
+     * named with the key "static/port_name". For example, if the
```
Use `\key` here. Same with the next few lines.

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -463,32 +487,125 @@ class SPROKIT_PIPELINE_EXPORT process
+     * If this port is not connected, the value supplied is taken from
+     * a specific config entry. The config entry is automatically
+     * named with the key "static/port_name". For example, if the
+     * port with this flag is named "foo", then the config entry for
```
foo → radius since we have an example below.

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -463,32 +487,125 @@ class SPROKIT_PIPELINE_EXPORT process
+          :static/foo  3.14159
+     * \endcode
+     *
+     * This flag may not be combined with \ref flag_required.
```
"…because it is implied that either the port must be connected or a static value be provided in the configuration."

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -463,32 +487,125 @@ class SPROKIT_PIPELINE_EXPORT process
+    /**
+     * \brief A flag which indicates that the port is required to be
+     * connected.
+     *
```
Extra space (and so on in this file).

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -525,12 +654,22 @@ class SPROKIT_PIPELINE_EXPORT process
     /**
      * \brief Method where subclass data processing occurs.
+     *
+     * In general, a process' _step method will involve:
```
Use `\c` for `_step`? Maybe another markup fits better though.

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -525,12 +654,22 @@ class SPROKIT_PIPELINE_EXPORT process
+     *
+     * In general, a process' _step method will involve:
+     *
+     * - Obtaining data from the input edges.
```
Use `\li` here.

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -525,12 +654,22 @@ class SPROKIT_PIPELINE_EXPORT process
      * \brief Runtime configuration for subclasses.
      *
+     * This method is called after the process is initially configured
+     * and satrted. A config block with updated values is supplied.
```
satrted → started

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -658,18 +799,39 @@ class SPROKIT_PIPELINE_EXPORT process
-     * \throws set_frequency_on_initialized_process_exception Thrown when the \p port's frequency is set after initialization.
+     * This method assigns a frequency to the input port. The number
+     * specifies how many inputs are to be accumulated between process
+     * step() calls. A frequency of one (the default) will give one
```
one → 1

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -658,18 +799,39 @@ class SPROKIT_PIPELINE_EXPORT process
      *
      * \param port The name of the port.
-     * \param new_frequency The frequency of the port.
+     * \param new_frequency The frequency of the port, in the range 1..n
```
There should be checks that the denominator is `1` somewhere (and tests I imagine).

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -658,18 +799,39 @@ class SPROKIT_PIPELINE_EXPORT process
+     * frequency of 4 will give the step() method 4 values in the
+     * queue for this input.
+     *
+     * A frequency of zero is a special case.
```
"Ports with a frequency of 0 are assumed to be non-regular and handled manually." Or something like that.

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -714,23 +877,51 @@ class SPROKIT_PIPELINE_EXPORT process
     /**
      * \brief Mark the process as complete.
+     *
+     * Calling this method within the step_() method indicates that
```
`step_` → `_step`

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -714,23 +877,51 @@ class SPROKIT_PIPELINE_EXPORT process
+     * the process has determined that it should not be called any
+     * more and that it is not going to produce any more data.
+     *
+     * It is considered good form to push a \c datum::complete_datum()
```
sprokit is an opinionated library. This can be more strongly worded to imply something much closer to "required".

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -714,23 +877,51 @@ class SPROKIT_PIPELINE_EXPORT process
      * \param port The port to get the edge for.
      *
-     * \return True if there is an edge connected to the \p port, or false if there is none.
+     * \return \b True if there is an edge connected to the \p port,
```
`\b`? Why not `\c` (same with `false` on the next line).

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -755,37 +946,63 @@ class SPROKIT_PIPELINE_EXPORT process
      * \brief Grab a datum from a port as a certain type.
      *
+     * This method grabs an input value directly from the port with no
+     * handling for static ports. The returned data value is undefined
```
Actually, this will block until one is available.

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -755,37 +946,63 @@ class SPROKIT_PIPELINE_EXPORT process
+     * behaves the same as grab_from_port_as().
+     *
+     * If there is no value at the port, then the value taken from the
+     * configuration entry "static/" + port_name is used.
```
`\key`

mathstuf commented on src/sprokit/pipeline/process.h:
```diff
@@ -755,37 +946,63 @@ class SPROKIT_PIPELINE_EXPORT process
+     *
+     * If the templated data type does not have a conversion from a
+     * string to an instance of the data type, you will need to
+     * provide one in the form of  an input operator.
```
This paragraph can be removed by mentioning `config_value` in the previous one. Something like "…is used from config_value()" maybe?

mathstuf commented on src/sprokit/pipeline/process_exception.h:
```diff
@@ -48,6 +48,9 @@
 {
 
+/// \todo need general runtime exception to report generic
+/// errors. Maybe style after std::exception hierarchy
```
I've just been using `std::runtime_error` for things that are outside the sprokit error domain (file read failure, file not existing, etc.).

mathstuf commented on src/sprokit/pipeline_util/doc/pipeline_declaration.dox:
```diff
@@ -75,6 +77,10 @@
  * \par Examples
  *
  * \todo Add some examples.
+*
```
Missing indent.

mathstuf commented on src/sprokit/pipeline_util/doc/pipeline_declaration.dox:
```diff
@@ -102,14 +108,14 @@
  *   <key-provider>      ::= <upper> <key-provider> | <upper>
  *   <key-provider-spec> ::= <provider-start> <key-provider> <provider-end>
- *   <key-options>       ::= <key-flag-decl> | <key-provider> | <key-flag-decl> <key-provider> | ""
+ *   <key-options>       ::= <key-flag-decl> | <key-provider-spec> | <key-flag-decl> <key-provider-spec> | ""
```
Good catch.

mathstuf commented on src/sprokit/pipeline_util/doc/pipeline_declaration.dox:
```diff
@@ -102,14 +108,14 @@
- * \todo Add some examples.
- *
  * <pre>
+ * :gsd{CONF} common.gsd
```
Should probably be "common:gsd".

mathstuf commented on src/sprokit/pipeline_util/doc/pipeline_declaration.dox:
```diff
@@ -102,14 +108,14 @@
- *
  * <pre>
+ * :gsd{CONF} common.gsd
+ * :tunable[ro]{ENV} TEST_ENV
```
Eh, "tunable" has a specific meaning in sprokit (runtime-changeable), so another name here would be better.

mathstuf commented on src/sprokit/pipeline_util/doc/pipeline_declaration.dox:
```diff
@@ -207,9 +227,28 @@
+   :static/<port-name> <key-value>
  * </pre>
+
+ * A port can not be connected and also have a static value.
```
Earlier docs state that the config value is ignored if the edge is connected. The code looks to support the other docs, so this should mention it as well.

mathstuf commented on src/sprokit/pipeline_util/doc/pipeline_declaration.dox:
```diff
@@ -268,4 +314,67 @@
  *
  * <pre>
  * </pre>
+
```
Missing leading `*`.

mathstuf commented on src/sprokit/pipeline_util/doc/pipeline_declaration.dox:
```diff
@@ -268,4 +314,67 @@
+ \subsection edge_config Edge Config
+
+Pipeline edges are configured in a hierarchical manner. First there is
+the _pipeline:_edge config block which establishes the basic
```
`\key`

mathstuf commented on src/sprokit/pipeline_util/doc/pipeline_declaration.dox:
```diff
@@ -268,4 +314,67 @@
+       :capacity 30
+</pre>
+
+Currently the only attribute that can be configured is "capacity".
```
`\key`

mathstuf commented on src/sprokit/pipeline_util/doc/pipeline_declaration.dox:
```diff
@@ -268,4 +314,67 @@
+        :<process>:<up_down>:<port>
+
+Where:
+<process> is the name of the process that is being connected.
```
A list would probably be better here.

mathstuf commented on src/sprokit/pipeline_util/doc/pipeline_declaration.dox:
```diff
@@ -268,4 +314,67 @@
+</pre>
+
+Both of these entries refer to the same edge, so in real life, you
+would only specify one.
```
specify → need. Though it might be good to mention that `down` overrides `up` configuration for an edge.

mathstuf commented on src/sprokit/pipeline_util/pipe_grammar.cxx:
```diff
@@ -399,6 +399,7 @@ ::common_grammar()
      |  qi::char_('-')
      |  qi::char_('_')
      |  qi::char_('/')
+     |  qi::char_('%') // temporaty solution to not allowing '.' as a key
```
Is there a reason `.` wouldn't work? `%` should at least have tests if allowed.

linus-sherrill replied:
This is an awful hack needed to support the requirement from OpenCV for '.'
in the key name. This was implemented in a hurry and is not seen as the
final solution. I tried to add a '.' to the grammar in place of the '%',
but an impenetrable runtime error was produced which I did not have time to
pursue. I am eagerly seeking alternate solutions.

On Fri, Aug 22, 2014 at 3:05 PM, Ben Boeckel notifications@github.com
wrote:

> In src/sprokit/pipeline_util/pipe_grammar.cxx:
> 
> > @@ -399,6 +399,7 @@ ::common_grammar()
> >       |  qi::char_('-')
> >       |  qi::char_('_')
> >       |  qi::char_('/')
> > -     |  qi::char_('%') // temporaty solution to not allowing '.' as a key
> 
> Is there a reason . wouldn't work? % should at least have tests if
> allowed.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Kitware/sprokit/pull/2/files#r16615426.

## 

*Linus Sherrill - *R&D Engineer
Kitware, Inc.
28 Corporate Drive
Clifton Park, NY 12065-8662
E: linus.sherrill@kitware.com
P: 518.881.4400

mathstuf replied:
Ah. It causes a grammar ambiguity in `proc.port` because there's a problem is down at the `port-addr` specification where `process_name` should replace `config_key`.

mathstuf replied:
Bleh, looks a little more complicated than that; some tests start failing with the "obvious" changes to the grammar parsing.

mathstuf commented on src/sprokit/tools/tool_usage.h:
```diff
@@ -44,7 +44,10 @@ SPROKIT_TOOLS_EXPORT void tool_version_message();
+SPROKIT_TOOLS_EXPORT boost::program_options::variables_map tool_parse(
+  int argc, char const* argv[],
+  boost::program_options::options_description const& desc,
+  std::string const& program_description= "" );
```
I don't see a reason to not make it required (you'll get a blank line anyways).

mathstuf commented on src/tools/pipe_config.cxx:
```diff
@@ -60,6 +60,29 @@
+// Description of this program and why I would want to use it
+static const std::string program_description(
+"This tool reads a pipeline configuration file and does some really\n"
+"cool stuff.  At its most basic, this tool will validate a pipeline\n"
```
Maybe a little too informal here :) . I've used it mainly for "compiling" a pipeline from all the includes and pieces and getting one of those documented files that `--output-config` for other tools dumped out.

linus-sherrill committed:
Addressed comments and issues

mathstuf committed:
Merge remote-tracking branch 'gh/linus-sherrill/kwiver-patch'

* gh/linus-sherrill/kwiver-patch:
  Convert config output operator to print() method usage.
  Documentation cleanup.
  Remove default for usage argument.
  Rename build target for all sprokit documentation.
  Convert output operator to print() method.
  Move demangle into utils
  Check for null pointer.
  Add missing include for release builds
  Add config output operator to simplify debugging.
  Add logging
  Add program description to tools infrastructure.
  Add program description to tools infrastructure.
  Add target to build combined documentation
  Documentation changes and add logging.

mathstuf committed:
Install the example Python scheduler

It's the only one right now that works with pipeline_runner as-is with
hybrid C++/Python pipelines.

linus-sherrill committed:
Added logger support.

Converted todo's and outputs to std::cerr to log message calls.

linus-sherrill committed:
Merge branch 'dev/kwiver-integration' of kwgitlab.kitware.com:computer-vision/sprokit into dev/kwiver-integration

# By Linus Sherrill (17) and Keith Fieldhouse (1)
* 'dev/kwiver-integration' of kwgitlab.kitware.com:computer-vision/sprokit:
  Addressed comments and issues
  Check for null pointer.
  Add missing include for release builds
  Improve documentation
  Expanding explanations
  Improved documentation, added examples
  Added logging and improved documentation
  Add config output operator to simplify debugging.
  Improve documentation
  Add logging
  Add program description to tools infrastructure.
  Documentation changes.
  Add program description to tools infrastructure.
  Documentation and error message improvements.
  Add target to build combined documentation
  Additional documentation
  Updated for newer doxygen program
  Updated Doxygen config file to newer version

Conflicts:
	src/sprokit/pipeline/edge.cxx
	src/sprokit/pipeline/modules.cxx
	src/sprokit/pipeline/pipeline.cxx
	src/sprokit/pipeline/process.h

linus-sherrill committed:
Convert source to use vital config_block
